central/credentialexpiry/service/service_impl.go

```diff
@@ -153,14 +162,82 @@ func (s *serviceImpl) getScannerCertExpiry(ctx context.Context) (*v1.GetCertExpi
 			errorList.AddError(err)
 			// All the endpoints have failed.
 			if len(errorList.ErrorStrings()) == len(clairifyEndpoints) {
-				return nil, errors.New(errorList.String())
+				return nil, errorList.ToError()
 			}
 		case expiry := <-expiryC:
 			return &v1.GetCertExpiry_Response{Expiry: expiry}, nil
 		}
 	}
 }
 
+func (s *serviceImpl) getScannerV4CertExpiry(ctx context.Context) (*v1.GetCertExpiry_Response, error) {
+	if !features.ScannerV4.Enabled() {
+		return nil, errors.Wrap(errox.InvalidArgs, "Scanner V4 is not integrated")
+	}
+	indexerConfig := s.scannerConfigs[mtls.ScannerV4IndexerSubject]
+	matcherConfig := s.scannerConfigs[mtls.ScannerV4MatcherSubject]
+	if indexerConfig == nil && matcherConfig == nil {
+		return nil, errors.New("could not load TLS configs to talk to Scanner V4 indexer and matcher")
+	}
+
+	integration, exists, err := s.imageIntegrations.GetImageIntegration(ctx, iiStore.DefaultScannerV4Integration.GetId())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve Scanner V4 image integration")
+	}
+	if !exists {
+		return nil, errors.New("Scanner V4 image integration missing")
+	}
+
+	s4Config := integration.GetScannerV4()
+	indexerEndpoint := scannerv4.DefaultIndexerEndpoint
+	if s4Config.GetIndexerEndpoint() != "" {
+		indexerEndpoint = s4Config.GetIndexerEndpoint()
+	}
+	matcherEndpoint := scannerv4.DefaultMatcherEndpoint
+	if s4Config.GetMatcherEndpoint() != "" {
+		matcherEndpoint = s4Config.GetMatcherEndpoint()
+	}
+
+	numEndpoints := 2
+	errC := make(chan error, numEndpoints)
+	expiryC := make(chan *types.Timestamp, numEndpoints)
+	getExpiry := func(subject mtls.Subject, endpoint string) {
+		expiry, err := s.getExpiryFunc(ctx, subject, s.scannerConfigs[subject], endpoint)
+		if err != nil {
+			errC <- err
+			return
+		}
+
+		log.Debugf("Obtained expiry for %q: %s", subject.Identifier, expiry)
+		expiryC <- expiry
+	}
+
+	go getExpiry(mtls.ScannerV4IndexerSubject, indexerEndpoint)
+	go getExpiry(mtls.ScannerV4MatcherSubject, matcherEndpoint)
+
+	errorList := errorhelpers.NewErrorList("failed to determine Scanner V4 cert expiry")
+	expiries := make([]*types.Timestamp, 0, numEndpoints)
+	for i := 0; i < numEndpoints; i++ {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case err := <-errC:
+			errorList.AddError(err)
+		case expiry := <-expiryC:
+			expiries = append(expiries, expiry)
+		}
+	}
+	if len(expiries) == 0 {
```

Comment:
would is also be an error if we only got one? I guess we only return one anyway, so perhaps that's fine. Only getting one probably indicates some other health-related problem we'd see elsewhere in the product

Reply:
The way it's built, no I didn't treat that as an error.  Reasoning was that its highly unlikely that indexer and matcher would have different expiration dates since the certs are always generated together - and the 'intent' of this API is not to determine the health of the dependant services, but to find out if certs are in good standing.